helper/querybuilder.php

```diff
@@ -0,0 +1,265 @@
+    protected $groupby;
+    /** @var string */
+    protected $having = '';
+    /** @var array */
+    protected $values = [];
+
+    /**
+     * Shorthand method: calls the appropriate getter deduced from $this->field
+     *
+     * @return array
+     */
+    public function getQuery()
+    {
+        if (!$this->field) {
+            throw new \RuntimeException('Failed to build a query, no field specified');
+        }
+        return ($this->field === 'pid') ? $this->getPages() : $this->getTags();
+    }
+
+    /**
+     * Processes all parts of the query for fetching tagged pages
+     *
+     * Returns the query builder object
+     *
+     * @return array
+     */
+    public function getPages()
+    {
+        $this->where = $this->getWhere();
+        $this->groupby = 'pid';
+        $this->orderby = "cnt DESC, pid";
+        if ($this->tags && $this->logicalAnd) $this->having = ' HAVING cnt = ' . count($this->tags);
+
+        return [$this->getSql(), $this->values];
+    }
+
+    /**
+     * Processes all parts of the query for fetching tags
+     *
+     * Returns the query builder object
+     *
+     * @return array
+     */
+    public function getTags()
+    {
+        $this->where = $this->getWhere();
+        $this->groupby = 'CLEANTAG(tag)';
+        $this->orderby = 'CLEANTAG(tag)';
+
+        return [$this->getSql(), $this->values];
+    }
+
+    /**
+     * Tags to search for
+     * @param array $tags
+     */
+    public function setTags($tags)
+    {
+        $this->tags = $tags;
+    }
+
+    /**
+     * Namespaces to limit search to
+     * @param array $ns
+     */
+    public function includeNS($ns)
+    {
+        $this->ns = $this->globNS($ns);
+    }
+
+    /**
+     * Namespaces to exclude from search
+     * @param array $ns
+     */
+    public function excludeNS($ns)
+    {
+        $this->notns = $this->globNS($ns);
+    }
+
+    /**
+     * Sets the logical operator used in tag search to AND
+     * @param bool $and
+     */
+    public function setLogicalAnd($and)
+    {
+        $this->logicalAnd = (bool)$and;
+    }
+
+    /**
+     * Result limit
+     * @param int $limit
+     */
+    public function setLimit($limit)
+    {
+        $this->limit = $limit;
+    }
+
+    /**
+     * Database field to select
+     * @param string $field
+     */
+    public function setField($field)
+    {
+        $this->field = $field;
+    }
+
+    /**
+     * Limit search to this page id
+     * @param string $pid
+     */
+    public function setPid($pid)
+    {
+        $this->pid = $pid;
+    }
+
+    /**
+     * Limit results to this tagger
+     * @param string $tagger
+     */
+    public function setTagger($tagger)
+    {
+        $this->tagger = $tagger;
+    }
+
+    /**
+     * Returns full query SQL
+     * @return string
+     */
+    protected function getSql()
+    {
+        $sql = "SELECT $this->field AS item, COUNT(*) AS cnt
+                  FROM taggings
+                 WHERE $this->where
```

Comment:
I don't think `$this->where` should exist. Instead you should call `$this->getWhere()` here to dynamically build the where clause when building the SQL.

Reply:
OK